[[ Android ]] Support additional android CPU architectures

livecodeian opened the pull request:
Allows for building against armv7, arm64, x86, x86_64

Commit:
Update config script to simplify cross-compilation

Various options useful for cross compilation have been added; use --help
to get more information.

Commit:
Use a host build of the server engine when cross-compiling

This is needed as we use the server engine to generate some of the
components of the build but, when cross-compiling, we can't run the server
engine for the target.

Commit:
Don't use CEF on non-x86(_64) Linux builds

For example, when targeting ARM.

Commit:
Update prebuilt scripts to support Linux cross-compilation

In particular, the ARMv6(HF) and ARMv7 targets.

Commit:
Add support for running tests under an interpreter

This is needed when running the test-suite for a cross-compiled engine: the
interpreter can be set to an emulator in order to execute the test suite
using the target executables.

Commit:
Add linux-armv6hf and linux-armv7 targets

Commit:
Fix a Gyp error in libGraphics

Gyp does not allow 'dependencies' blocks inside a 'target_conditions'
block; this change moves the conditions out.

Commit:
Update the prebuilt scripts to support multiple Android archs

Because Android is based on Linux, the Linux support gets an update too.

Commit:
Add Linux/Android arm64 support to foundation.h

Commit:
Don't redefine socklen_t on Android

Its system headers already define it.

Commit:
Don't include <sys/time64.h> for 64-bit Android builds

The time support is already 64-bit and this header does not work.

Commit:
Explicitly link to libatomic in Android builds

If the engine is compiled in Thumb mode for ARMv6, libatomic is needed to
support atomic operations. For ARMv7, ARM64, x86 and x86_64 the library is
not required but linking to it is harmless.

Commit:
Don't try to build cefprocess for non Windows or Linux targets

Commit:
Ensure JNI types get defined properly in 64-bit Android builds

Sometimes, we'd end up with jlong defined as 'long' and other times as
'long long' which causes linking failures. This seems to depend on the
order of inclusion of various headers.

Commit:
Fix the specification of the Android library path

The case of an empty library path (e.g. when using a standalone Android
toolchain) was not handled properly and could cause strange linker errors.

Commit:
Add support to our configure scripts for more Android architectures

The script now recognises armv7, arm64, x86 and x86_64 architectures in
addition to the existing armv6 architecture.

Commit:
Make the Android linker script architecture-agnostic

livecodeian committed:
[[ Android ]] Fix gyp problems on macosx
Make revbrowser-process target & dependencies conditional on platform support for libcef

livecodeian committed:
[[ Android ]] Add platform triple entries for android architectures armv7, arm64, x86, x86_64

livecodeian committed:
[[ Android ]] Don't attempt to download prebuilt archives for not yet supported linux targets

livecodeian committed:
Merge remote-tracking branch 'origin/develop-9.0' into feature-android-armv7

livecodeian committed:
[[ Android ]] Update API target for additional android archs

livecodeian committed:
Merge branch 'develop-9.0' of https://github.com/livecode/livecode into feature-android-armv7

livecodeali commented:
@livecode-vulcan review ok 32c0658

livecodeali commented:
@livecode-vulcan review ok 32c0658

livecode-vulcan commented:
💙 review by @livecodeali ok 32c0658ada56635a3f6d9988f8fe9f442cb8c4fd

livecode-vulcan commented:
😞 test failure 32c0658ada56635a3f6d9988f8fe9f442cb8c4fd

* try-community-armv6-android-sdk26_ndk16r15: [pending](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/13)
* try-community-armv7-android-sdk26_ndk16r15: [pending](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/5)
* try-community-arm64-android-sdk26_ndk16r15: [pending](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/5)
* try-community-x86-android-sdk26_ndk16r15: [pending](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/5)
* try-community-x86_64-android-sdk26_ndk16r15: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/5)
* try-community-js-emscripten-sdk1.35: [failure](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/1527)
* try-community-universal-ios-iphoneos11.4: [failure](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos11.4/builds/54)
* try-community-universal-ios-iphonesimulator11.4: [failure](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator11.4/builds/54)
* try-community-universal-mac-macosx10.9: [failure](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/1557)
* try-community-x86-linux-debian8: [failure](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/1551)
* try-community-x86_64-linux-debian8: [failure](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/1547)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/1584)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/1524)

livecodeali commented on prebuilt/libopenssl.gyp:
```diff
@@ -117,7 +121,7 @@
 					[
-						'OS == "android"',
+						'toolset_os == "android"',
 						{
```
There needs to be an additional 
```
 							'conditions':
 							[
 								[
 									'OS == "android"',
 									{
```
here like with libicu.gyp, otherwise the configure step complains about `<(android_subplatform)`

livecodeian committed:
[[ Android ]] Work-around issues in gyp handling of paths in target_conditions block

livecodeali commented:
Travis has this error:
```
make[2]: *** No rule to make target `../../prebuilt/lib/linux/x86_64/CEF/natives_blob.bin', needed by `out/Debug/natives_blob.bin'.  Stop.
make[2]: Leaving directory `/home/travis/build/livecode/livecode/build-linux-x86_64/livecode'
make[1]: *** [compile-linux-x86_64] Error 2
make[1]: Leaving directory `/home/travis/build/livecode/livecode'
make: *** [all-linux-x86_64] Error 2
```

livecodeian committed:
[[ Android ]] Remove check for CEF platform support on Linux

*Note* A corrected version of this check will need to be implemented
when additional Linux architecture support is added.

livecodeali commented:
@livecode-vulcan review ok 44fa19f

livecode-vulcan commented:
💙 review by @livecodeali ok 44fa19fd5f7c98a031d1d7f2b01d024e90466c95

livecode-vulcan commented:
😞 test failure 44fa19fd5f7c98a031d1d7f2b01d024e90466c95

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/14)
* try-community-armv7-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/6)
* try-community-arm64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/6)
* try-community-x86-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/6)
* try-community-x86_64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/6)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/1528)
* try-community-universal-ios-iphoneos11.4: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos11.4/builds/55)
* try-community-universal-ios-iphonesimulator11.4: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator11.4/builds/55)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/1558)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/1552)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/1548)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/1585)
* try-community-x86_64-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/1525)

livecodeali commented:
@livecodeian Not sure if there is a CEF problem with windows 64 bit here or not - will retry and see what happens

livecodeali commented:
@livecode-vulcan retry 44fa19f

livecode-vulcan commented:
😞 test failure 44fa19fd5f7c98a031d1d7f2b01d024e90466c95

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/25)
* try-community-armv7-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/17)
* try-community-arm64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/17)
* try-community-x86-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/17)
* try-community-x86_64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/17)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/1539)
* try-community-universal-ios-iphoneos11.4: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos11.4/builds/66)
* try-community-universal-ios-iphonesimulator11.4: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator11.4/builds/66)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/1569)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/1563)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/1559)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/1596)
* try-community-x86_64-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/1536)

Commit:
Merge branch 'develop-9.0' of https://github.com/livecode/livecode into feature-android-armv7

montegoulding commented:
@livecode-vulcan retry 44fa19f

livecode-vulcan commented:
😞 test failure 44fa19fd5f7c98a031d1d7f2b01d024e90466c95

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/32)
* try-community-armv7-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/24)
* try-community-arm64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/24)
* try-community-x86-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/24)
* try-community-x86_64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/24)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/1546)
* try-community-universal-ios-iphoneos11.4: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos11.4/builds/73)
* try-community-universal-ios-iphonesimulator11.4: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator11.4/builds/73)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/1576)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/1570)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/1566)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/1603)
* try-community-x86_64-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/1543)

livecodeali commented:
@livecodeian There's still an issue here with 64 bit windows and CEF it seems...

Commit:
[[ Android ]] Use uniform_arch when locating windows x86_64 CEF libs

livecodeali commented:
@livecode-vulcan review ok b0e4458

livecode-vulcan commented:
💙 review by @livecodeali ok b0e4458b308b8d4fba9ee61e5d6173e50b51fbc5

livecode-vulcan commented:
😎 test success b0e4458b308b8d4fba9ee61e5d6173e50b51fbc5

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/35)
* try-community-armv7-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/27)
* try-community-arm64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/27)
* try-community-x86-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/27)
* try-community-x86_64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/27)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/1549)
* try-community-universal-ios-iphoneos11.4: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos11.4/builds/76)
* try-community-universal-ios-iphonesimulator11.4: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator11.4/builds/76)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/1579)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/1573)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/1569)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/1606)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/1546)
